### ISSUE TITLE
Support Json format for `dotnet new list` 

### ISSUE THREAD

**Issue by MoaidHathot** [dotnet/sdk#43679]
### Is your feature request related to a problem? Please describe.
Calling `dotnet new list` generates a table-based output. Due to terminal size constraints some values may be automatically truncated, making it very hard for other tools to use the .NET CLI (like a [Neovim extension](https://github.com/MoaidHathot/dotnet.nvim)).

### Describe the solution you'd like
Support `--format json` as it already supported in other commands.

### Additional context
![image](https://github.com/user-attachments/assets/6d8d4090-882e-4370-a342-b3f36930e615)

**Comment by baronfel** [comment]
This is definitely a good thing to do, and very in line with our more recently-created commands. We will likely take a converted stab at making sure there is machine readable output for all of our commands for the next release or two, but if this is especially important to you we'd love to take a PR adding the flag to 'dotnet new list'.

**Comment by nagilson** [comment]
@MoaidHathot Do you plan on making a PR for this? Want to make sure if you did any work we don't discount any contribution you might be working on if we decide to pick this up.

**Comment by MoaidHathot** [comment]
Hi @nagilson, yes, I'm working on a PR. Planning to publish it or a draft of it in the following couple of days.

**Comment by mariuszbyahoo** [comment]
@MoaidHathot please confirm that you're working on the task - I've recently finished cooperation with my employer and I would use some C# tasks to do in my spare time (which I have freaaaaaking lot of it!)

**Comment by MoaidHathot** [comment]
Hi @baronfel, @nagilson, @mariuszbyahoo, I've created the following draft PR https://github.com/dotnet/sdk/pull/44140/files and I'm seeking review.
I created the draft so that it is easier to discuss the design before making more complete changes.
The thing I'm not sure about is whether we should reuse `TemplateGroupDisplay` and `TemplateGroupTableRow` as I've done, although they are part of the `TabularOutput` namespace. We can move them from the `TabularOutput`.

The following is the output I'm currently getting, as a PoC, we will probably change it a bit:
```json
[
  {
    "Author": "Microsoft",
    "Classifications": "MSBuild/props",
    "Languages": "",
    "Name": "MSBuild Directory.Build.props file",
    "ShortNames": "buildprops",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "MSBuild/props",
    "Languages": "",
    "Name": "MSBuild Directory.Build.targets file",
    "ShortNames": "buildtargets",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "dotnet gitattributes file",
    "ShortNames": "gitattributes,.gitattributes",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "dotnet gitignore file",
    "ShortNames": "gitignore,.gitignore",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "global.json file",
    "ShortNames": "globaljson,global.json",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "NuGet Config",
    "ShortNames": "nugetconfig,nuget.config",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "MSBuild/packages/props/CPM",
    "Languages": "",
    "Name": "MSBuild Directory.Packages.props file",
    "ShortNames": "packagesprops",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Solution",
    "Languages": "",
    "Name": "Solution File",
    "ShortNames": "sln,solution",
    "Type": ""
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "Dotnet local tool manifest file",
    "ShortNames": "tool-manifest",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "Web Config",
    "ShortNames": "webconfig",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/Library",
    "Languages": "[C#],F#,VB",
    "Name": "Class Library",
    "ShortNames": "classlib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/Console",
    "Languages": "[C#],F#,VB",
    "Name": "Console App",
    "ShortNames": "console",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Test/MSTest",
    "Languages": "[C#],F#,VB",
    "Name": "MSTest Test Class",
    "ShortNames": "mstest-class",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Test/MSTest",
    "Languages": "[C#],F#,VB",
    "Name": "MSTest Test Project",
    "ShortNames": "mstest",
    "Type": "project"
  },
  {
    "Author": "Aleksei Kharlov aka halex2005 (codeofclimber.ru)",
    "Classifications": "Test/NUnit",
    "Languages": "[C#],F#,VB",
    "Name": "NUnit 3 Test Item",
    "ShortNames": "nunit-test",
    "Type": "item"
  },
  {
    "Author": "Aleksei Kharlov aka halex2005 (codeofclimber.ru)",
    "Classifications": "Test/NUnit",
    "Languages": "[C#],F#,VB",
    "Name": "NUnit 3 Test Project",
    "ShortNames": "nunit",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Test/MSTest/Playwright",
    "Languages": "[C#]",
    "Name": "MSTest Playwright Test Project",
    "ShortNames": "mstest-playwright",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Test/NUnit/Playwright",
    "Languages": "[C#]",
    "Name": "NUnit Playwright Test Project",
    "ShortNames": "nunit-playwright",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Test/xUnit",
    "Languages": "[C#],F#,VB",
    "Name": "xUnit Test Project",
    "ShortNames": "xunit",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "API Controller",
    "ShortNames": "apicontroller",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "MVC Controller",
    "ShortNames": "mvccontroller",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/gRPC",
    "Languages": "",
    "Name": "Protocol Buffer File",
    "ShortNames": "proto",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "Razor Component",
    "ShortNames": "razorcomponent",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "Razor Page",
    "ShortNames": "page",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "Razor View",
    "ShortNames": "view",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "MVC ViewImports",
    "ShortNames": "viewimports",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/ASP.NET",
    "Languages": "[C#]",
    "Name": "MVC ViewStart",
    "ShortNames": "viewstart",
    "Type": "item"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/Worker/Web",
    "Languages": "[C#],F#",
    "Name": "Worker Service",
    "ShortNames": "worker",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Blazor/WebAssembly",
    "Languages": "[C#]",
    "Name": "Blazor Web App",
    "ShortNames": "blazor",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Blazor/WebAssembly/PWA",
    "Languages": "[C#]",
    "Name": "Blazor WebAssembly Standalone App",
    "ShortNames": "blazorwasm",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Empty",
    "Languages": "[C#],F#",
    "Name": "ASP.NET Core Empty",
    "ShortNames": "web",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/gRPC/API/Service",
    "Languages": "[C#]",
    "Name": "ASP.NET Core gRPC Service",
    "ShortNames": "grpc",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Razor/Library",
    "Languages": "[C#]",
    "Name": "Razor Class Library",
    "ShortNames": "razorclasslib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/MVC/Razor Pages",
    "Languages": "[C#]",
    "Name": "ASP.NET Core Web App (Razor Pages)",
    "ShortNames": "webapp,razor",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/MVC",
    "Languages": "[C#],F#",
    "Name": "ASP.NET Core Web App (Model-View-Controller)",
    "ShortNames": "mvc",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Web API/API/Service",
    "Languages": "[C#],F#",
    "Name": "ASP.NET Core Web API",
    "ShortNames": "webapi",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Web/Web API/API/Service",
    "Languages": "[C#]",
    "Name": "ASP.NET Core Web API (native AOT)",
    "ShortNames": "webapiaot",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WinForms",
    "Languages": "[C#],VB",
    "Name": "Windows Forms App",
    "ShortNames": "winforms",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WinForms",
    "Languages": "[C#],VB",
    "Name": "Windows Forms Control Library",
    "ShortNames": "winformscontrollib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WinForms",
    "Languages": "[C#],VB",
    "Name": "Windows Forms Class Library",
    "ShortNames": "winformslib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WPF",
    "Languages": "[C#],VB",
    "Name": "WPF Application",
    "ShortNames": "wpf",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WPF",
    "Languages": "[C#],VB",
    "Name": "WPF Class Library",
    "ShortNames": "wpflib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WPF",
    "Languages": "[C#],VB",
    "Name": "WPF Custom Control Library",
    "ShortNames": "wpfcustomcontrollib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Common/WPF",
    "Languages": "[C#],VB",
    "Name": "WPF User Control Library",
    "ShortNames": "wpfusercontrollib",
    "Type": "project"
  },
  {
    "Author": "Microsoft",
    "Classifications": "Config",
    "Languages": "",
    "Name": "EditorConfig file",
    "ShortNames": "editorconfig,.editorconfig",
    "Type": "item"
  }
]
```

**Comment by baronfel** [comment]
@MoaidHathot thanks for putting out the draft! A few notes on the format from my end:

* When you say "we will probably change it a bit so that the output is not an array, but a json object containing a property of type array", what's driving that thought?
* There are a few aspects of the current `TemplateGroupTableRow` that I would want to change in a purely-JSON representation:
  * `Classifications` should be an array of the string tags, not `/`-delimited
  * `Languages` should be an array of `{ id: string, default: bool }` objects to capture the semantics of the list of languages and which one is currently the one that will be chosen absent any user override
  * `ShortNames` should be an array of the names, not `,`-separated

Basically, the json should be as idiomatic JSON as possible. I'd love for this to be an actual class/structure in the codebase that we can serialize an array of (or something) so that we have an actual contract that we have to think about evolving if we want to change it.

**Comment by MoaidHathot** [comment]
@baronfel, thank you for the quick review.

> Basically, the json should be as idiomatic JSON as possible. I'd love for this to be an actual class/structure in the codebase that we can serialize an array of (or something) so that we have an actual contract that we have to think about evolving if we want to change it.

Totally agree, I'll make the change

> When you say "we will probably change it a bit so that the output is not an array, but a json object containing a property of type array", what's driving that thought?

Since it seems to be the standard in other `--format json` results, although all or most of them are ForwardedApp Commands to the Nuget CLI.

For example, the following is the output of `dotnet list .\src\Dumpify\Dumpify.csproj package --format json`.
It contain a schema `version` which I think is important.
```json
{
  "version": 1,
  "parameters": "",
  "problems": [
    {
      "level": "warning",
      "text": "(A) : Auto-referenced package."
    }
  ],
  "projects": [
    {
      "path": "P:/Github/Dumpify/src/Dumpify/Dumpify.csproj",
      "frameworks": [
        {
          "framework": "netstandard2.0",
          "topLevelPackages": [
            {
              "id": "NETStandard.Library",
              "requestedVersion": "[2.0.3, )",
              "resolvedVersion": "2.0.3",
              "autoReferenced": "true"
            },
            {
              "id": "Spectre.Console",
              "requestedVersion": "0.48.0",
              "resolvedVersion": "0.48.0"
            }
          ]
        },
        {
          "framework": "netstandard2.1",
          "topLevelPackages": [
            {
              "id": "Spectre.Console",
              "requestedVersion": "0.48.0",
              "resolvedVersion": "0.48.0"
            }
          ]
        },
        {
          "framework": "net8.0",
          "topLevelPackages": [
            {
              "id": "Spectre.Console",
              "requestedVersion": "0.48.0",
              "resolvedVersion": "0.48.0"
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by baronfel** [comment]
Ok, that's what I expected - you're thinking about protocol extensibility and such. That makes perfect sense, just wanted to make sure what I interpreted your meaning as was what you meant your meaning to be interpreted as :)
